command/agent/template/template.go

```diff
@@ -0,0 +1,186 @@
+	"github.com/hashicorp/vault/command/agent/config"
+)
+
+// ServerConfig is a config struct for setting up the basic parts of the
+// Server
+type ServerConfig struct {
+	Logger hclog.Logger
+	// Client        *api.Client
+	VaultConf     *config.Vault
+	ExitAfterAuth bool
+
+	Namespace string
+}
+
+// Server manages the Consul Template Runner which renders templates
+type Server struct {
+	// config holds the ServerConfig used to create it. It's passed along in other
+	// methods
+	config *ServerConfig
+
+	// runner is the consul-template runner
+	runner *manager.Runner
+
+	// unblockCh is used to block until a template is rendered
+	unblockCh chan struct{}
+
+	// Templates holds the parsed Consul Templates
+	Templates []*ctconfig.TemplateConfig
+
+	DoneCh        chan struct{}
+	logger        hclog.Logger
+	exitAfterAuth bool
+}
+
+// NewServer returns a new configured server
+func NewServer(conf *ServerConfig) (*Server, chan struct{}) {
+	unblock := make(chan struct{})
+	ts := Server{
+		DoneCh:        make(chan struct{}),
+		logger:        conf.Logger,
+		unblockCh:     unblock,
+		config:        conf,
+		exitAfterAuth: conf.ExitAfterAuth,
+	}
+	return &ts, unblock
+}
+
+// Run kicks off the internal Consul Template runner, and listens for changes to
+// the token from the AuthHandler. If Done() is called on the context, shut down
+// the Runner and return
+func (ts *Server) Run(ctx context.Context, incoming chan string, templates []*ctconfig.TemplateConfig) {
+	latestToken := new(string)
+	ts.logger.Info("starting sink server")
+	defer func() {
+		ts.logger.Info("template server stopped")
+		close(ts.DoneCh)
+	}()
+
+	if incoming == nil {
+		panic("incoming channel is nil")
+	}
+
+	// If there are no templates, close the unblockCh
+	if len(templates) == 0 {
+		// nothing to do
+		close(ts.unblockCh)
+		return
+	}
+
+	// construct a consul template vault config based the agents vault
+	// configuration
+	var runnerConfig *ctconfig.Config
+	if runnerConfig = newRunnerConfig(ts.config, templates); runnerConfig == nil {
+		ts.logger.Info("template server failed to generate runner config")
+		return
+	}
+
+	var err error
+	ts.runner, err = manager.NewRunner(runnerConfig, false)
+	if err != nil {
+		ts.logger.Info("template server failed to create")
+		return
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			ts.runner.StopImmediately()
+			return
+
+		case token := <-incoming:
+			if token != *latestToken {
+				ts.logger.Info("template server received new token")
+				ts.runner.Stop()
+				*latestToken = token
+				ctv := ctconfig.Config{
+					Vault: &ctconfig.VaultConfig{
+						Token: latestToken,
+					},
+				}
+				runnerConfig.Merge(&ctv)
+				runnerConfig.Finalize()
+				var runnerErr error
+				ts.runner, runnerErr = manager.NewRunner(runnerConfig, false)
+				if runnerErr != nil {
+					ts.logger.Info("template server failed with new Vault token")
+					// TODO: continue or fail here? Right now we just continue
```

Comment:
That's a good question. Is a panic the only way to fail? Also probably should at least be warn level logging.

Reply:
Definitely open to ideas here. I'm imagining a situation where policies changed and the token doesn't have permissions to access the secret, though maybe that wouldn't surface here. I'll dig in more to the possible errors scenarios. 

Regarding failing, no, panic is not the only way, we could shut everything down and return. That would leave agent running but no templates would be rendered. 

Regarding the logging; I changed that to an `Error` and included the error in 14daed773